plugin/clone/src/clone_common.cc

```diff
@@ -0,0 +1,114 @@
+/*
+  Copyright (C) 2022, 2023, Laurynas Biveinis
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; version 2 of the License.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */
+
+#include "plugin/clone/include/clone_common.h"
+
+#include <string>
+#include <utility>
+
+#include "lex_string.h"
+#include "my_dbug.h"
+#include "my_inttypes.h"
+#include "plugin/clone/include/clone_hton.h"
+#include "sql/handler.h"
+#include "sql/json_dom.h"
+#include "sql/sql_plugin_ref.h"
+#include "storage/perfschema/table_log_status.h"
+
+namespace myclone {
+
+// Call precopy for the participating storage engines that have the
+// precopy method.
+int Ha_clone_common_cbk::precopy(THD *thd, uint task_id) {
+  const auto &all_locators = get_all_locators();
+
+  // Must be called by InnoDB the clone-driving engine
+  assert(all_locators[0].m_hton->db_type == DB_TYPE_INNODB);
+  assert(get_loc_index() == 0);
+  auto *const saved_cbk_hton = get_hton();
+
+  uint index = 0;
+  for (const auto &loc : all_locators) {
+    auto *const precopy_fn = loc.m_hton->clone_interface.clone_precopy;
+    if (precopy_fn) {
+      set_loc_index(index);
+      const auto res =
+          precopy_fn(loc.m_hton, thd, loc.m_loc, loc.m_loc_len, task_id, this);
+      if (res != 0) {
+        set_loc_index(0);
+        set_hton(saved_cbk_hton);
+        return res;
+      }
+    }
+    ++index;
+  }
+  set_loc_index(0);
+  set_hton(saved_cbk_hton);
+
+  return 0;
+}
+
+// Perform the cross-engine synchronization: execute a
+// performance_schema.log_status query, and call set_log_stop for each storage
+// engine with its part of STORAGE_ENGINES column JSON object from that query.
+int Ha_clone_common_cbk::synchronize_engines() {
+  const auto &all_locators = get_all_locators();
+
+  std::unique_ptr<table_log_status> table{static_cast<table_log_status *>(
+      table_log_status::create(&table_log_status::m_share))};
+
+  auto err = table->rnd_init(true);
+  assert(err == 0);
+
+  err = table->rnd_next();
```

Comment:
> To clarify, if fetching the log status information fails is this how the server would fail out of the clone process? 

Yes

> Given we've seen failures occasional, should we retry this synchronization call?

I would try to wait to collect more instances of this failure to inform the restart policy here, if possible. If the failure is caused by a RocksDB error, then only an `EINTR` could be handled in a reasonable way, but my guess (didn't check) that it should be handled by RocksDB internally already. As for anything else, I could speculate that e.g. an intermittent I/O error is a sign of more serious underlying issues anyway.